hive/src/main/java/org/apache/iceberg/hive/legacy/HiveExpressions.java

```diff
@@ -0,0 +1,322 @@
+
+  private HiveExpressions() {}
+
+  /**
+   * Simplifies the {@link Expression} so that it fits the restrictions of the expression that can be passed
+   * to the Hive metastore. For details about the simplification, please see {@link RemoveNonPartitionPredicates} and
+   * {@link RewriteUnsupportedOperators}
+   * @param expr The {@link Expression} to be simplified
+   * @param partitionColumnNames The set of partition column names
+   * @return TRUE if the simplified expression results in an always true expression or if there are no predicates on
+   *            partition columns in the simplified expression,
+   *         FALSE if the simplified expression results in an always false expression,
+   *         otherwise returns the simplified expression
+   */
+  static Expression simplifyPartitionFilter(Expression expr, Set<String> partitionColumnNames) {
+    try {
+      Expression partitionPredicatesOnly = ExpressionVisitors.visit(expr,
+          new RemoveNonPartitionPredicates(partitionColumnNames));
+      if (partitionPredicatesOnly == null) {
+        return Expressions.alwaysTrue();
+      } else {
+        return ExpressionVisitors.visit(partitionPredicatesOnly, new RewriteUnsupportedOperators());
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Error while processing expression: " + expr, e);
+    }
+  }
+
+  /**
+   * Converts an {@link Expression} into a filter string which can be passed to the Hive metastore
+   * @param expr The {@link Expression} to be converted into a filter string. This expression must fit the restrictions
+   *             on Hive metastore partition filters. For more details, see {@link RewriteUnsupportedOperators}
+   * @return a filter string equivalent to the given {@link Expression} which can be passed to the Hive metastore
+   */
+  static String toPartitionFilterString(Expression expr) {
+    return ExpressionVisitors.visit(expr, ExpressionToPartitionFilterString.get());
+  }
+
+  /**
+   * Removes any predicates on non-partition columns from the given {@link Expression}
+   */
+  private static class RemoveNonPartitionPredicates extends ExpressionVisitors.ExpressionVisitor<Expression> {
+
+    private final Set<String> partitionColumnNamesLowerCase;
+
+    RemoveNonPartitionPredicates(Set<String> partitionColumnNames) {
+      this.partitionColumnNamesLowerCase =
+          partitionColumnNames.stream().map(String::toLowerCase).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Expression alwaysTrue() {
+      return Expressions.alwaysTrue();
+    }
+
+    @Override
+    public Expression alwaysFalse() {
+      return Expressions.alwaysFalse();
+    }
+
+    @Override
+    public Expression not(Expression result) {
+      return (result == null) ? null : Expressions.not(result);
+    }
+
+    @Override
+    public Expression and(Expression leftResult, Expression rightResult) {
+      if (leftResult == null && rightResult == null) {
+        return null;
```

Comment:
What are the consequences of substituting non-partition predicates with alwaysTrue()? Seems it will simplify the code?

Reply:
Also, it may clear confusions about the semantics of null.

Reply:
Yes, seems like this can be simplified. I had kept it null previously to distinguish actual TRUE expressions passed by user v/s ones added by visitor. But it makes sense to just substitute non-partition predicates with alwaysTrue()